### ISSUE TITLE
0.7.0.dev1/dev2: Qwen3.8-Flash-Next quants emit immediate EOS instead of tool calls in assistant-less contexts (0.6.4 clean, byte-identical A/B)

### ISSUE THREAD

**Issue by layer4down** [jundot/omlx#3660]
## Summary

Since 0.7.0.dev1, Qwen3.8-Flash-Next oQ quants (reproduced primarily on `Qwen3.8-Flash-Next-oQ5e-mtp`, once on an oQ4e build) emit **immediate EOS instead of tool calls** in long assistant-less contexts. The final message is whitespace-only (`"\n\n"`), `finish_reason: "stop"`, zero `tool_calls`, while the reasoning channel ends mid-plan (e.g. reasoning ends with "I'll read the config:" and then the text channel emits whitespace and stops). 0.6.4 handles the byte-identical requests correctly. **0.7.0.dev2 does not fix it — it's worse** (details below).

## Environment

- oMLX builds: 0.6.4 (`260830015308`) ✅ clean · 0.7.0.dev1 (`260911022938`) ❌ · 0.7.0.dev2 (`260911232929`) ❌ (build numbers from `Resources/omlx/_build_info.py`)
- Apple M2 Ultra 192GB, macOS 26.x
- Model: `Qwen3.8-Flash-Next-oQ5e-mtp` (community GBP-DE build; single weights file used across all three engines)
- Settings: temperature 1.0 (also affected at other temps), MTP/Lightning OFF, ANE OFF, DFlash OFF, SpecPrefill OFF, TurboQuant KV OFF, thinking on

## The A/B (byte-identical requests, only the engine differs)

The exact same streaming `/v1/chat/completions` request — same weights file, same system prompt, same message array, same params — replayed against each version:

| Version | Trials | Outcome |
|---|---|---|
| 0.6.4 | 5 | **0 EOS — all 5 emit `finish_reason: "tool_calls"`** (model engages the code tool) |
| 0.7.0.dev1 | 10 | **10 EOS** — `stop` / 0 `tool_calls` / whitespace-only content |
| 0.7.0.dev2 | 5 | **5 EOS** — identical signature |

Timeline corroboration: after a server restart onto dev1, the first occurrence appeared **9 minutes later**, with zero occurrences across weeks on pre-0.7 engines including the same quant on agentic workloads.

## Reproduction (synthetic — no private content needed)

The trigger shape is a **large system prompt followed by an assistant-less user wall** (many consecutive `user` messages, no `assistant`/tool messages between them) with a code-execution tool defined. A synthetic generator reproduces it:

```python
import json, requests

sys_prompt = ("You are a software engineering agent with filesystem and shell tools. "
              "Be systematic and verify on disk. ") * 900          # ~124k chars
messages = [{"role": "system", "content": sys_prompt}]
for i in range(19):
    messages.append({"role": "user", "content": f"Task note {i}: please continue the investigation."})

body = {
    "model": "Qwen3.8-Flash-Next-oQ5e-mtp:low-qwen38",
    "messages": messages,
    "tools": [{"type": "function", "function": {
        "name": "run_code", "description": "Run a shell command and return output",
        "parameters": {"type": "object", "properties": {"cmd": {"type": "string"}},
                        "required": ["cmd"]}}}],
    "stream": True,
    "max_completion_tokens": 32768,
}
r = requests.post("http://127.0.0.1:8000/v1/chat/completions", json=body, stream=True)
# join deltas; on 0.7.0.dev1/dev2: reasoning plans tool use, then content == "\n\n", stop, no tool_calls
```

- 0.6.4: model emits a `run_code` call.
- 0.7.0.dev1: reasoning ends mid-plan → whitespace-only content → stop.

Simpler second shape (agentic clients): after any **completed** turn, send a minimal user message like `"."` — the next assistant message frequently collapses to `"\n\n"`+stop with no tool call.

## 0.7.0.dev2 is worse

On dev2 the collapse no longer needs the wall. A fresh, short agentic session (simple "review these files" task) degenerated: 4 consecutive assistant messages of pure whitespace across 3 steps, **zero tool calls while the reasoning channel plans file reads at every step**. A corrective user nudge also EOS'd (on dev1 the same nudge recovers a full answer).

## Suspected mechanism

PR #3520 (0.7.0.dev1): *"Gathered QSA attention now also accelerates eligible text decode… greedy outputs can differ from dense attention."* With MTP/ANE/DFlash/SpecPrefill all off, gathered-QSA is the decode approximation these models do use, and there is **no settings toggle** for it. The regression window (0.6.4 → 0.7.0.dev1) and the greedy-decode character of the failure both point there; #3469's decode rewrite is the secondary candidate. For precedent, 0.6.4's own #3279 fixed an earlier "immediate EOS" class on Qwen3.8 paths, so this class has had engine-side fixes before.

## Ask

1. A toggle (or fix) for the gathered-QSA decode approximation so greedy tool-calling output can be restored to dense-attention behavior.
2. If the cause is elsewhere in the 0.7.0 decode rewrite, the A/B above (same request bytes across the three builds) should bisect it quickly.

Happy to run any additional bisect between the two build ranges if useful.

**Comment by jundot** [comment]
Thanks @layer4down for the detailed reproduction. I tested the synthetic prompt on an M3 Ultra with an oQ4e checkpoint, both with and without the native QSA kernels. The original ~17k-token prompt produced tool calls under both greedy decoding and the default sampling settings, and an extended ~34k-token prompt also completed with tool calls. With the native kernels enabled, I verified that the QSA indexer, top-k, and sparse attention paths were actually used.

Enabling the kernels changed the generated text under greedy decoding, but neither configuration produced the whitespace-only stop with no tool calls. I also tried sending `.` after a completed response with the kernels enabled and got normal text responses.

I wouldn’t expect the M3/M5 difference alone to explain this, but I’ll run additional tests on M5 to check. My checkpoint is oQ4e rather than your primary oQ5e reproduction, so that difference remains to be checked as well.

**Comment by layer4down** [comment]
Slight correction: M2 Ultra 192GB is my host platform. Got my Flash Next agent testing fixes. Will submit anything we find that might be useful. Thanks.

EDIT: also this happens with oQ4e and oQ5e, all reasoning_effort levels.

**Comment by layer4down** [comment]
**Agent disclosure:** this comment (and the analysis behind it) was prepared by an AI agent acting for @layer4down, with their review and authorization: **GLM-5.3 (Z.ai)** running in **Claude Code v2.1.271**. The investigation on our side was itself largely conducted by an agent using the same **Qwen3.8-Flash-Next oQ5e-mtp** weights (community GBP-DE build) served locally by the installed **oMLX 0.6.4** (build `260830015308`), with a source build of **0.7.0.dev2** as the device-under-test. We'd rather be explicit about the tooling than fuzzy about it.

**Correction to the report:** our host is an **Apple M2 Ultra 192GB**, not "M5 Studio" — the Environment line above has been fixed. Apologies; that typo likely seeded the hardware-variance thread.

**What we found: the whitespace-stop is a server-side silent drop, not a model or silicon issue.**

On 0.7.0.dev1/dev2, when the model names a function that is **not in the request's `tools` array**, the qwen3_coder streaming path loses the entire turn:

1. The model emits a **complete, well-formed** `<tool_call>…</tool_call><|im_end|>` envelope — verified with an in-process hook on the model's forward pass; logits are sane and every envelope token is top-1 at emission.
2. `ToolCallStreamFilter` suppresses the envelope from content deltas (by design).
3. The early-stream gate introduced in **#3327** (`registered_tool_names` check in `stream_chat_completion`) drops the **successfully parsed** call via a bare `continue` when its name isn't registered.
4. Terminal recovery can't rescue it: mlx_lm's function regex anchors `</function>` at end-of-string, and `_marker_payloads` only scans strictly nested envelopes — a trailing `<|im_end|>` after the close marker defeats both.

Client-visible result: `finish_reason: "stop"`, zero `tool_calls`, zero content — exactly this issue's signature. The gate is absent from 0.6.4 (`git tag --contains` puts #3327 in dev1/dev2 only), matching the version matrix in the report.

**Why your reproduction misses it — and why M5 testing won't hit it either:** prompts that keep the model in-grammar (registered tool names) never trigger the gate. It keys on **prompt shape**, not length or platform: contexts that induce an unregistered function name. In our case, assistant-less user-wall histories where the model reaches for a pseudo-tool — e.g. emitting `<function=skill>` while only `run_code` is registered — reproduce 27/27. Your oQ4e runs on M3 Ultra are consistent with this: different checkpoint, in-grammar prompts, gate never fires.

**Candidate fix (happy to PR):** at the rejection branch, yield the captured envelope back as a visible content delta instead of `continue`-ing. This honors `ToolCallStreamFilter`'s own documented contract ("a malformed attempt can be returned intact instead of partially hidden"), mirrors the existing K2 parsing-failure fallback, and the #1830/#1854 no-silent-drop philosophy already in `parse_tool_calls`. Validation on the failing shape: 5/5 replays that previously ended silent-empty now stream the envelope as content (byte-identical to the instrumented token trace); registered-name calls unregressed (3/3); AST regression test included.

One honesty note: we separately observe long-context stalls on 0.6.4 itself at ~90k cached tokens — a different signature still under characterization; the fix above does not claim to address that.
